### ISSUE TITLE
Could we add an option to disable the "Select a Kit" dialog box?

### ISSUE THREAD

**Issue by ilg-ul** [microsoft/vscode-cmake-tools#2538]
### Brief Issue Summary

Kits and presets are probably a great way of managing toolchains for regular projects, but for multi-platform projects that have their own way of keeping track of toolchains (possibly different for different build configurations), the dialog box asking to "Select a Kit" when opening the project is not only of no value, but even annoying.

### Additional Information

Here is an example when opening such a project:

![Screenshot 2022-05-11 at 21 00 34](https://user-images.githubusercontent.com/362732/167916508-3acaad7f-6950-4d61-a258-eec4196fb10c.png)

This project keeps track of all dependencies, including toolchains, in [`package.json`](https://github.com/micro-os-plus/utils-lists-xpack/blob/xpack/package.json).

Such projects are managed in VS Code with the [xPack C/C++ Managed Build Tools](https://marketplace.visualstudio.com/items?itemName=ilg-vscode.xpack) extension, which is based on the C/C++ extension, and is neutral to the build system.

The xPack extension uses neither kits nor presets. It would be great to have a project configuration option to disable any user interaction asking for kits or presets when they are not needed.

If this change is not very complicated, and you provide some hints where to start, I can try to submit a PR.

**Comment by ilg-ul** [comment]
After taking a second look at the xPack extension sources, I think that the only functionality needed from CMake Tools is the `CMakeLists.txt` editor. 

So, the question is how to allow projects that include `CMakeList.txt` files but are managed by other extensions, to run without interferences from the CMake Tools extension, but still benefit from the editor.

**Comment by bobbrow** [comment]
We don't actually have any editor features for CMakeLists.txt right now.  They should be coming from the other CMake extension that we bundle with ours.

We could potentially add a setting for this scenario similar to the one you helped us with yesterday though.

**Comment by ilg-ul** [comment]
> We don't actually have any editor features for CMakeLists.txt right now. They should be coming from the other CMake extension that we bundle with ours.

Ah, right, the editor comes from the separate CMake extension, by twxs. I did a test, I disabled the CMake Tools entirely and re-enable the CMake extension, and the editor was functional.

> We could potentially add a setting for this scenario similar to the one you helped us with yesterday though.

Ok. In this case how about a configuration option that prevents the CMake Tools extension activation entirely?

**Comment by bobbrow** [comment]
> In this case how about a configuration option that prevents the CMake Tools extension activation entirely? Or allows the activation, but disables most of the functionality?

That option already exists in VS Code.  Go to the Extensions panel and right click on the entry for CMake Tools.  You will get the option to disable the extension globally or for the current workspace.  You can also just uninstall the extension.

![image](https://user-images.githubusercontent.com/12818240/167938198-f326ec42-1afd-4e4c-ba46-5cc2d276212f.png)

**Comment by ilg-ul** [comment]
> Go to the Extensions panel and right click on the entry for CMake Tools. You will get the option to disable the extension globally or for the current workspace. You can also just uninstall the extension.

I checked these options, but they are not very convenient. CMake Tools is the mainstream CMake extension, and I expect most of my users to have it installed, and use it successfully for their other projects. I don't want to interfere in any way with their use case, and definitely I cannot ask them to uninstall the extension.

The xPack extension is a specific use case intended mainly for embedded projects, and I see no problem to be used in parallel with CMake Tools for other projects.

Thus a more convenient approach would be to have a configuration option in the `setting.json` of the projects that use the xPack extension, to disable some of the functionality of CMake Tools that is not needed.

At first I identified the dialog box asking for kits, but there might be other such undesired interactions with the user, that I do not know, or that might be added in the future.

I would appreciate your suggestion on what exactly to disable, since the code is quite complex.

**Comment by bobbrow** [comment]
I don't think it's a good idea for us to go carve up the extension into a set of settings that enable/disable bits and pieces.  Those settings can be abused by other extensions, effectively disabling us without the user's consent.

The "select a kit" dropdown appears only once per workspace if you don't dismiss it (e.g. pick the "Unspecified" kit) and doesn't show at all if you are using CMake Presets.

**Comment by ilg-ul** [comment]
> The "select a kit" dropdown appears only once per workspace

I'm in the process of splitting a large mono-repo project into separate libraries. For now there are more than 30 of them. This means more than 30 useless dropdowns to deal with. You have no idea how annoying this can be.

**Comment by ilg-ul** [comment]
> The "select a kit" dropdown ... doesn't show at all if you are using CMake Presets

I tried to add a `CMakePresets.json`. The "select a kit" dropdown did not appear, but then a "Select a configure preset" dropdown appeared. I'm afraid this route is not very helpful.

**Comment by ilg-ul** [comment]
> We could potentially add a setting for this scenario similar to the one you helped us with yesterday though.

If disabling activation is not a good idea, I reverted to the initial proposal to disable only the dropdown, taking a route similar to the previous contribution, as you suggested.

I checked the result on my projects and it seems functional.

**Comment by ilg-ul** [comment]
If I understand your proposal right, in the PR you suggested to extend the functionality of `"configureOnOpen": false` and add some extra checks to skip undesired functionality.

This is fine with me, especially since after silencing _Select a Kit_ we also have to silence the _CMake Tools would like to configure IntelliSense for this folder."_ dropdown.

![Screenshot 2022-05-19 at 08 53 35](https://user-images.githubusercontent.com/362732/169222359-389319c1-ca25-455d-8ef4-d6692d189f45.png)

**Comment by bobbrow** [comment]
Yes, I do think that configureOnOpen: false is probably the right way to go.  Sorry I haven't had time to do additional testing/research in the last week.  I was hoping you might play around with it and see if it will meet your needs.  If not, maybe we can schedule a call and talk about what your goals are.  It feels like your goals center around disabling and replacing our extension's functionality, and if that's the case I would still be of the opinion that your users should choose one extension or the other and we shouldn't be adding significant amounts of code to disable ourselves.

**Comment by ilg-ul** [comment]
> I would still be of the opinion that your users should choose one extension or the other 

I think this is an unfair proposal to the users.

The scenarios is the following: users have multiple projects, some use traditional CMake, some use the new xPack build framework. The xPack framework is an extension of traditional CMake, providing a neutral solution for multi-platform and multi-configuration. It is true that the xPack extension can run without the CMake Tools extension, but I don't think that it is for me to impose this on the users, if they need CMake Tools for their other projects, they should be allowed to have both installed in parallel.

As for your previous proposal to disable the CMake Tools extension specifically for a given workspace, if you know of a setting that I can add to my projects to do this automatically without bothering the user with annoying questions, please let me know and I can try it.

> I was hoping you might play around with it and see if it will meet your needs. If not, maybe we can schedule a call and talk about what your goals are.

Ah, I thought that you are playing around and suggest a more detailed solution. In this case I'll try to take a look and see if I find a solution. We can schedule a Skype call at any time, if you think useful.

The goal is to have both extensions installed, allow CMake traditional projects to use the CMake Tools extension without any limitations, but for projects that use the xPack extension, allow the user to enable one or more settings to silence the dropdowns that are not relevant. I don't know how many features of CMake Tools would someone want to use with xPack projects, and I would not be inclined to make myself the decision to disable the CMake Tools extension, the choice should be made by the user.

**Comment by ilg-ul** [comment]
BTW, the philosophical issue (pertaining to many extensions, it is not a specific critique to CMake Tools) is the policy to automatically try to monopolise the user interface and try to take full control of the workflow, based on some conditions, like the presence of some files, **without giving the user a choice to opt out**. 

In this case the presence of a CMakeLists.txt file is a good condition, but the user should be the ultimate choice maker if the extension must be activated or not. 

Disabling the extension for a specific workspace it far from perfect, since the setting is stored outside the project and is tied to the folder path; thus projects freshly cloned require manual intervention each time, and moving the project to a different folder requires repeated manual intervention to disable the extension again.

**Comment by bobbrow** [comment]
I don't see any problem with the current philosophy.  Asking the user to decide which extensions to enable every time they open a new folder doesn't scale because users install tons of extensions.  The activationEvents exist precisely for this reason.  It may come across as "monopolization," but I still think that the user has the controls they need to enable/disable extensions based on their needs in each workspace.

I'm not saying that the current method is perfect and has no room for improvement.  I'm just saying that if you want to advocate for change here, the best place to do it would be in the VS Code repo where the team who has control of these things can engage in the conversation.  Even with the changes we're discussing here, your users are still going to see the popups the first time they load any new folder until they make the settings changes or disable the extension through the VS Code UI.  I consider both methods to be more or less equivalent.

**Comment by ilg-ul** [comment]
> Asking the user to decide which extensions to enable every time they open a new folder doesn't scale because users install tons of extensions.

This is not what I suggested, this is exactly what I'm advocating agains, asking the user irelevant questions.

> Even with the changes we're discussing here, your users are still going to see the popups the first time they load any new folder until they make the settings changes or disable the extension through the VS Code UI. 

Not really. The plan is to enable these options in all my xPack projects, and also in the templates used to generate xPack projects, so the default will be set properly and the user will not be asked a thing. Plus that the user, if necessary, will still be able to change the option in settings.json at any time.

> I consider both methods to be more or less equivalent.

Disabling an extension for a workspace is not equivalent with setting an option in settings.json, since the later can be set when the project is created, it is committed in Git and remains persistent, regardless of where and how many times the project is downloaded or moved.

Perhaps we should schedule a call, to give you a better understanding of the use case.

**Comment by ilg-ul** [comment]
I would still appreciate if we can schedule a call to talk directly. When are you available?

**Comment by bobbrow** [comment]
I'll send you a message and we can coordinate over email.

**Comment by bobbrow** [comment]
1.12.11 (pre-release) was published this morning. Give it a try and see if setting `"cmake.configureOnOpen": false` does what you want now.

**Comment by ilg-ul** [comment]
After setting the option and renaming the project folder to avoid possible cached configs, there were no more dialog boxes when i opened the project in VS Code, so I would say that the patch was effective.

Thank you very much!

**Comment by ilg-ul** [comment]
I wonder if this option cannot be extended to also disable the question asked in #2532? Right now I have to add two configuration options to my projects to silence the CMake extension, which is a bit more tedious than necessary.

**Comment by ilg-ul** [comment]
@bobbrow, the fix you implemented disabled the kit selection dialog, but I noticed that although `configureOnOpen` is false, in some cases the CMake extension continuously flashes a small error in the bottom right corner with the following message:

```
Bad CMake executable: "". Check to make sure it is installed or the value of the "cmake.cmakePath" setting contains the correct path
```

The message is right, there is no CMake in the path and the `cmake.cmakePath` is not set, since the xPack extension has its own path management.

The error pops while editing, I would say that after autosave.

Can we do something to get rid of this error? Do you want me to open a new issue for this?

**Comment by bobbrow** [comment]
> Can we do something to get rid of this error? Do you want me to open a new issue for this?

Yes, please open a new issue for this.

**Comment by ilg-ul** [comment]
Done, as #2666.
